SIKFromAddress data race

Looks like there is a data race, probably the SIK tree is updated while, at the same time, a Get operation is performed.
When the tree is updated, the nodes on the tree change and those that are no longer required, are removed. If a Get operation is in course, it will fail at some point because some nodes of the Path do not exist anymore.

```
2023-11-22T14:00:34.636Z DBG middleware/logger.go:158 > "GET https://api-dev.vocdoni.net:443/metrics HTTP/2.0" from 95.217.216.185:54082 - 200 21252B in 14.135206ms
2023-11-22T14:00:43.088Z INF ipfs storage addresses=26934 peers=405 pins=2289
2023-11-22T14:00:43.81Z INF offchain downloader enqueued=0 retrying=0 total=447
2023-11-22T14:00:49.977Z INF vochain status blockTime="1m:10.00 10m:10.34 1h:10.40 6h:10.40 24h:10.40" elections=579 height=110053 mempool=0 peers=4 voteCache=0 votes=4883 votes/min=0
2023-11-22T14:00:50.702Z INF subpub network cluster=1 known=26957 peers=213
2023-11-22T14:00:51.371Z DBG ipfs/ipfs.go:314 > retrieved file /ipfs/bafybeigwsmihkf7krxkfuqejflfrao43g3ortsw4i6r6ouyzljp7mlci4e from cache
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x98 pc=0x2f38533]

goroutine 125009948 [running]:
github.com/cockroachdb/pebble.(*Batch).Get(0xc0175f2a48?, {0xc01d15ae10?, 0x0?, 0x0?})
	github.com/cockroachdb/pebble@v0.0.0-20230928194634-aa077af62593/batch.go:550 +0x13
go.vocdoni.io/dvote/db/pebbledb.get({0x4507ea0?, 0x0?}, {0xc01d15ae10?, 0x366d320?, 0x1?})
	go.vocdoni.io/dvote/db/pebbledb/pebledb.go:21 +0x33
go.vocdoni.io/dvote/db/pebbledb.(*WriteTx).Get(0x0?, {0xc01d15ae10?, 0x0?, 0x0?})
	go.vocdoni.io/dvote/db/pebbledb/pebledb.go:69 +0x2e
go.vocdoni.io/dvote/db/prefixeddb.(*PrefixedWriteTx).Get(0xc02635d4a0, {0xc00b77df12, 0x20, 0x0?})
	go.vocdoni.io/dvote/db/prefixeddb/prefixeddb.go:117 +0x10c
go.vocdoni.io/dvote/tree/arbo.(*Tree).down(0xc00f5358c0, {0x7f7f64cee2f8, 0xc02635d4a0}, {0xc015c6a330, 0x14, 0x14}, {0xc00b77df12, 0x20, 0x20}, {0xc016896c00, ...}, ...)
	go.vocdoni.io/dvote/tree/arbo/navigate.go:24 +0xee
go.vocdoni.io/dvote/tree/arbo.(*Tree).down(0xc00f5358c0, {0x7f7f64cee2f8, 0xc02635d4a0}, {0xc015c6a330, 0x14, 0x14}, {0xc00b77dea2, 0x20, 0x40}, {0xc016896c00, ...}, ...)
	go.vocdoni.io/dvote/tree/arbo/navigate.go:77 +0x851
go.vocdoni.io/dvote/tree/arbo.(*Tree).down(0xc00f5358c0, {0x7f7f64cee2f8, 0xc02635d4a0}, {0xc015c6a330, 0x14, 0x14}, {0xc00b77de72, 0x20, 0x20}, {0xc01f150ea0, ...}, ...)
	go.vocdoni.io/dvote/tree/arbo/navigate.go:82 +0xaaf
go.vocdoni.io/dvote/tree/arbo.(*Tree).down(0xc00f5358c0, {0x7f7f64cee2f8, 0xc02635d4a0}, {0xc015c6a330, 0x14, 0x14}, {0xc00b77ddd2, 0x20, 0x20}, {0xc01f150ea0, ...}, ...)
	go.vocdoni.io/dvote/tree/arbo/navigate.go:77 +0x851
go.vocdoni.io/dvote/tree/arbo.(*Tree).down(0xc00f5358c0, {0x7f7f64cee2f8, 0xc02635d4a0}, {0xc015c6a330, 0x14, 0x14}, {0xc00b77dd82, 0x20, 0x20}, {0xc00316c2a0, ...}, ...)
	go.vocdoni.io/dvote/tree/arbo/navigate.go:77 +0x851
go.vocdoni.io/dvote/tree/arbo.(*Tree).down(0xc00f5358c0, {0x7f7f64cee2f8, 0xc02635d4a0}, {0xc015c6a330, 0x14, 0x14}, {0xc00b77dd12, 0x20, 0x40}, {0xc02040d008, ...}, ...)
	go.vocdoni.io/dvote/tree/arbo/navigate.go:77 +0x851
go.vocdoni.io/dvote/tree/arbo.(*Tree).down(0xc00f5358c0, {0x7f7f64cee2f8, 0xc02635d4a0}, {0xc015c6a330, 0x14, 0x14}, {0xc016135f60, 0x20, 0x20}, {0x0, ...}, ...)
	go.vocdoni.io/dvote/tree/arbo/navigate.go:82 +0xaaf
go.vocdoni.io/dvote/tree/arbo.(*Tree).GetWithTx(0xc00f5358c0, {0x7f7f64cee2f8, 0xc02635d4a0}, {0xc015c6a330, 0x14, 0x14})
	go.vocdoni.io/dvote/tree/arbo/tree.go:541 +0x453
go.vocdoni.io/dvote/tree.(*Tree).Get(0x13c2e1d?, {0x7f7f64cee2f8?, 0xc02635d4a0?}, {0xc015c6a330?, 0x36cdd40?, 0x1c000230638?})
	go.vocdoni.io/dvote/tree/tree.go:70 +0x77
go.vocdoni.io/dvote/statedb.(*TreeUpdate).Get(...)
	go.vocdoni.io/dvote/statedb/treeupdate.go:63
go.vocdoni.io/dvote/statedb.(*TreeUpdate).DeepGet(0xc00316c000?, {0xc015c6a330, 0x14, 0x14}, {0xc00316c000?, 0x36cdd40?, 0xc00066a601?})
	go.vocdoni.io/dvote/statedb/treeupdate.go:192 +0x8f
go.vocdoni.io/dvote/statedb.(*treeUpdateView).DeepGet(0x3c94224?, {0xc015c6a330?, 0x0?, 0xab?}, {0xc00316c000?, 0xc016ed0a00?, 0xc0175f3d08?})
	go.vocdoni.io/dvote/statedb/treeupdate.go:378 +0x1d
go.vocdoni.io/dvote/vochain/state.(*State).SIKFromAddress(0x3631ac0?, {0x2e, 0x5b, 0x2, 0x23, 0x7f, 0x44, 0x89, 0xd5, 0x3a, ...})
	go.vocdoni.io/dvote/vochain/state/sik.go:46 +0x145
go.vocdoni.io/dvote/api.(*API).accountHandler(0xc00f7696c0, 0x25cd77b?, 0xc016bf0b60)
	go.vocdoni.io/dvote/api/accounts.go:150 +0x4b8
go.vocdoni.io/dvote/httprouter/apirest.(*API).RegisterMethod.func1({{0x35ed6e0, 0xc01644eb70}, {0xc14fa044d614e10a, 0x50020feb9464, 0x5fbd0a0}, {0xc00cffa390, 0x3, 0x3}, 0xc016bf0b60})
	go.vocdoni.io/dvote/httprouter/apirest/apirest.go:229 +0x6b
created by go.vocdoni.io/dvote/httprouter.(*HTTProuter).AddPublicHandler.(*HTTProuter).routerHandler.func1 in goroutine 125010181
	go.vocdoni.io/dvote/httprouter/httprouter.go:283 +0x3c5
```

--- a/vochain/state/sik.go
+++ b/vochain/state/sik.go
@@ -43,13 +43,13 @@ type SIK []byte
 // SIKFromAddress function return the current SIK value associated to the provided
 // address.
 func (v *State) SIKFromAddress(address common.Address) (SIK, error) {
+	v.tx.RLock()
+	defer v.tx.RUnlock()
+
 	siksTree, err := v.tx.DeepSubTree(StateTreeCfg(TreeSIK))
 	if err != nil {
 		return nil, fmt.Errorf("%w: %w", ErrSIKSubTree, err)
 	}
-	v.tx.Lock()
-	defer v.tx.Unlock()
-
 	sik, err := siksTree.Get(address.Bytes())
 	if err != nil {
 		if errors.Is(err, arbo.ErrKeyNotFound) {
@@ -69,24 +69,21 @@ func (v *State) SIKFromAddress(address common.Address) (SIK, error) {
 //   - If it exists but it is not valid, overwrite the stored value with the
 //     provided one.
 func (v *State) SetAddressSIK(address common.Address, newSIK SIK) error {
+	v.tx.Lock()
+	defer v.tx.Unlock()
 	siksTree, err := v.tx.DeepSubTree(StateTreeCfg(TreeSIK))
 	if err != nil {
 		return fmt.Errorf("%w: %w", ErrSIKSubTree, err)
 	}
 	// check if exists a registered sik for the provided address, query also for
 	// no committed tree version
-	v.tx.Lock()
 	rawSIK, err := siksTree.Get(address.Bytes())
-	v.tx.Unlock()
 	if errors.Is(err, arbo.ErrKeyNotFound) {
 		// if not exists create it
 		log.Debugw("setSIK (create)",
 			"address", address.String(),
 			"sik", newSIK.String())
-		v.tx.Lock()
-		err = siksTree.Add(address.Bytes(), newSIK)
-		v.tx.Unlock()
-		if err != nil {
+		if err := siksTree.Add(address.Bytes(), newSIK); err != nil {
 			return fmt.Errorf("%w: %w", ErrSIKSet, err)
 		}
 		return nil
@@ -102,10 +99,7 @@ func (v *State) SetAddressSIK(address common.Address, newSIK SIK) error {
 		"address", address.String(),
 		"sik", SIK(rawSIK).String())
 	// if the hysteresis is reached update the sik for the address
-	v.tx.Lock()
-	err = siksTree.Set(address.Bytes(), newSIK)
-	v.tx.Unlock()
-	if err != nil {
+	if err := siksTree.Set(address.Bytes(), newSIK); err != nil {
 		return fmt.Errorf("%w: %w", ErrSIKSet, err)
 	}
 	return nil
@@ -117,14 +111,14 @@ func (v *State) SetAddressSIK(address common.Address, newSIK SIK) error {
 // prevent it from being updated until all processes created before that height
 // have finished.
 func (v *State) InvalidateSIK(address common.Address) error {
+	v.tx.Lock()
+	defer v.tx.Unlock()
 	siksTree, err := v.tx.DeepSubTree(StateTreeCfg(TreeSIK))
 	if err != nil {
 		return fmt.Errorf("%w: %w", ErrSIKSubTree, err)
 	}
 	// if the sik does not exists or something fails querying return the error
-	v.tx.Lock()
 	rawSIK, err := siksTree.Get(address.Bytes())
-	v.tx.Unlock()
 	if err != nil {
 		return fmt.Errorf("%w: %w", ErrSIKGet, err)
 	}
@@ -132,11 +126,8 @@ func (v *State) InvalidateSIK(address common.Address) error {
 	if !SIK(rawSIK).Valid() {
 		return ErrSIKAlreadyInvalid
 	}
-	v.tx.Lock()
 	invalidatedSIK := make(SIK, sikLeafValueLen).InvalidateAt(v.CurrentHeight())
-	err = siksTree.Set(address.Bytes(), invalidatedSIK)
-	v.tx.Unlock()
-	if err != nil {
+	if err := siksTree.Set(address.Bytes(), invalidatedSIK); err != nil {
 		return fmt.Errorf("%w: %w", ErrSIKDelete, err)
 	}
 	return nil
@@ -200,6 +191,8 @@ func (v *State) UpdateSIKRoots() error {
 	currentBlock := v.CurrentHeight()
 
 	// get sik roots key-value database associated to the siks tree
+	v.tx.RLock()
+	defer v.tx.RUnlock()
 	siksTree, err := v.tx.DeepSubTree(StateTreeCfg(TreeSIK))
 	if err != nil {
 		return fmt.Errorf("%w: %w", ErrSIKSubTree, err)
@@ -383,8 +376,8 @@ func (v *State) PurgeSIKsByElection(pid []byte) error {
 // SIKGenProof returns the proof of the provided address in the SIKs tree.
 // The first returned value is the leaf value and the second the proof siblings.
 func (v *State) SIKGenProof(address common.Address) ([]byte, []byte, error) {
-	v.tx.Lock()
-	defer v.tx.Unlock()
+	v.tx.RLock()
+	defer v.tx.RUnlock()
 	siksTree, err := v.tx.DeepSubTree(StateTreeCfg(TreeSIK))
 	if err != nil {
 		return nil, nil, fmt.Errorf("%w: %w", ErrSIKSubTree, err)
@@ -395,8 +388,8 @@ func (v *State) SIKGenProof(address common.Address) ([]byte, []byte, error) {
 
 // SIKRoot returns the last root hash of the SIK merkle tree.
 func (v *State) SIKRoot() ([]byte, error) {
-	v.tx.Lock()
-	defer v.tx.Unlock()
+	v.tx.RLock()
+	defer v.tx.RUnlock()
 	siksTree, err := v.tx.DeepSubTree(StateTreeCfg(TreeSIK))
 	if err != nil {
 		v.tx.Unlock()

--- a/vochain/state/sik.go
+++ b/vochain/state/sik.go
@@ -43,7 +43,14 @@ type SIK []byte
 // SIKFromAddress function return the current SIK value associated to the provided
 // address.
 func (v *State) SIKFromAddress(address common.Address) (SIK, error) {
-	sik, err := v.mainTreeViewer(false).DeepGet(address.Bytes(), StateTreeCfg(TreeSIK))
+	siksTree, err := v.tx.DeepSubTree(StateTreeCfg(TreeSIK))
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", ErrSIKSubTree, err)
+	}
+	v.tx.Lock()
+	defer v.tx.Unlock()
+
+	sik, err := siksTree.Get(address.Bytes())
 	if err != nil {
 		if errors.Is(err, arbo.ErrKeyNotFound) {
 			return nil, fmt.Errorf("%w: %w", ErrSIKNotFound, err)
@@ -110,8 +117,14 @@ func (v *State) SetAddressSIK(address common.Address, newSIK SIK) error {
 // prevent it from being updated until all processes created before that height
 // have finished.
 func (v *State) InvalidateSIK(address common.Address) error {
+	siksTree, err := v.tx.DeepSubTree(StateTreeCfg(TreeSIK))
+	if err != nil {
+		return fmt.Errorf("%w: %w", ErrSIKSubTree, err)
+	}
 	// if the sik does not exists or something fails querying return the error
-	rawSIK, err := v.mainTreeViewer(false).DeepGet(address.Bytes(), StateTreeCfg(TreeSIK))
+	v.tx.Lock()
+	rawSIK, err := siksTree.Get(address.Bytes())
+	v.tx.Unlock()
 	if err != nil {
 		return fmt.Errorf("%w: %w", ErrSIKGet, err)
 	}
@@ -121,7 +134,7 @@ func (v *State) InvalidateSIK(address common.Address) error {
 	}
 	v.tx.Lock()
 	invalidatedSIK := make(SIK, sikLeafValueLen).InvalidateAt(v.CurrentHeight())
-	err = v.tx.DeepSet(address.Bytes(), invalidatedSIK, StateTreeCfg(TreeSIK))
+	err = siksTree.Set(address.Bytes(), invalidatedSIK)
 	v.tx.Unlock()
 	if err != nil {
 		return fmt.Errorf("%w: %w", ErrSIKDelete, err)

--- a/vochain/state/sik_test.go
+++ b/vochain/state/sik_test.go
@@ -3,6 +3,8 @@ package state
 import (
 	"bytes"
 	"encoding/hex"
+	"errors"
+	"sync"
 	"testing"
 
 	"github.com/ethereum/go-ethereum/common"
@@ -204,4 +206,52 @@ func Test_validSIK(t *testing.T) {
 
 	input, _ = hex.DecodeString("F3668000B66c61aAa08aBC559a8C78Ae7E007C2e")
 	qt.Assert(t, SIK(input).Valid(), qt.IsTrue)
+}
+
+func TestSIKDataRace(t *testing.T) {
+	c := qt.New(t)
+	// create a state for testing
+	dir := t.TempDir()
+	s, err := NewState(db.TypePebble, dir)
+	qt.Assert(t, err, qt.IsNil)
+
+	// create some siks
+	addrs := []common.Address{}
+	siks := map[common.Address]SIK{}
+	for i := 0; i < 10000; i++ {
+		s := ethereum.NewSignKeys()
+		c.Assert(s.Generate(), qt.IsNil)
+		sik, err := s.AccountSIK(nil)
+		c.Assert(err, qt.IsNil)
+		addrs = append(addrs, s.Address())
+		siks[s.Address()] = sik
+	}
+
+	wg := &sync.WaitGroup{}
+	iterations := len(addrs) * 10
+
+	wg.Add(2)
+	go func() {
+		defer wg.Done()
+
+		for i := 0; i < iterations; i++ {
+			idx := util.RandomInt(0, len(addrs)-1)
+			addr := addrs[idx]
+			if err := s.SetAddressSIK(addr, siks[addr]); err != nil {
+				c.Assert(errors.Is(err, ErrRegisteredValidSIK), qt.IsTrue)
+			}
+		}
+	}()
+	go func() {
+		defer wg.Done()
+
+		idx := util.RandomInt(0, len(addrs)-1)
+		addr := addrs[idx]
+		if _, err := s.SIKFromAddress(addr); err != nil {
+			c.Assert(errors.Is(err, ErrSIKNotFound), qt.IsTrue)
+			return
+		}
+		c.Assert(s.InvalidateSIK(addr), qt.IsNil)
+	}()
+	wg.Wait()
 }

--- a/vochain/state/state.go
+++ b/vochain/state/state.go
@@ -348,12 +348,12 @@ func (v *State) Save() ([]byte, error) {
 	// Commit the statedb tx
 	// Note that we need to commit the tx after calling listeners, because
 	// the listeners may need to get the previous (not committed) state.
-	v.tx.Lock()
-	defer v.tx.Unlock()
 	// Update the SIK merkle-tree roots
 	if err := v.UpdateSIKRoots(); err != nil {
 		return nil, fmt.Errorf("cannot update SIK roots: %w", err)
 	}
+	v.tx.Lock()
+	defer v.tx.Unlock()
 	err := func() error {
 		var err error
 		if err := v.tx.SaveWithoutCommit(); err != nil {

